### ISSUE TITLE
Line-Classes with multiple Styles not displayed correctly when using attributes

### ISSUE THREAD

**Issue by krone-rvr** [MapServer/MapServer#6933]
## Expected behavior and actual behavior.

So we are currently trying to reduce our mapfile size by using clever features such as the scale token and stuff.
While styling our road network, we encountered strange behavior when a class containing two overlapping styles that are definded by attributes is rendered.

We are loading data via a postgis connection

```
DATA "geometry from (select *, NULL as nullid from 
	(
		SELECT 
		geometry,
		type,
		%width_motorway_in% AS width_motorway_in,
		%width_motorway_out% AS width_motorway_out
		FROM web.roads
	) as roads
	) as nullidq using unique nullid using srid=25832"
```
	
Then defining two scale token objects

motorway outline:

    SCALETOKEN
		NAME "%width_motorway_out%"
		VALUES
			"0"		"300.0"
			"125"		"200.0"
			"250"		"100.0"
			...
			"750000"	"2.0"
			"1500000"	"1.5"
		END
	END
  
motorway inline:

	SCALETOKEN
		NAME "%width_motorway_in%"
		VALUES
			"0"		"296.0"
			"125"		"196.0"
			"250"		"97.0"
			...
			"750000"	"1.5"
			"1500000"	"1.0"
		END
	END
	
And finally setting up the road class by defining two styles to display motorways with an outline

	CLASS
		EXPRESSION ('[type]' = 'motorway')
		STYLE
			WIDTH [width_motorway_out]
			COLOR "#969696" 
			MAXWIDTH 300
		END 
		STYLE
			WIDTH [width_motorway_in]
			COLOR "#ffbe5a"     
			MAXWIDTH 300
		END
	END
	
Instead of displaying our street data as a whole, each object is rendered on its own.

<img width="1387" alt="actual" src="https://github.com/MapServer/MapServer/assets/129192217/9765e61a-b710-419f-903a-92b331c405f0">

When changing the class definition and hardcoding the values that are otherwise provided by our scale tokens, the streets are displayed as expected.

	CLASS
		EXPRESSION ('[type]' = 'motorway')
		STYLE
			WIDTH 22
			COLOR "#969696" 
			MAXWIDTH 300
		END 
		STYLE
			WIDTH 20
			COLOR "#ffbe5a"     
			MAXWIDTH 300
		END
	END
	
<img width="1385" alt="exppected" src="https://github.com/MapServer/MapServer/assets/129192217/feb3d701-a30e-49e8-9b33-f15e35f4681c">

Our current workaround is creating two separate layer objects for street in- and outlines respectively, but that seems a bit cumbersome and is what we are actually trying to avoid.

This behavior has occured with two completely different datasets on both MapServer Version 7.6.5 and 8.0.1 on Debian GNU/Linux 10 (buster)

**Comment by geographika** [comment]
@krone-rvr - I can recreate this in https://app.mapserverstudio.net/#SgOu2oEV

Interestingly the following classes producesthe same effect as using scaletokens (with each line having a border), with hard-coded widths. It looks like when SCALETOKENs are used it takes the same rendering approach.

```
        CLASS
            STYLE
                WIDTH 8
                COLOR "#969696"
                MAXWIDTH 300
            END
        END
        CLASS
            STYLE
                WIDTH 6
                COLOR "#ffbe5a"
                MAXWIDTH 300
            END
        END
```

![image](https://github.com/MapServer/MapServer/assets/490840/5db75d57-7cd7-4397-b596-1e8338010a5f)

**Comment by geographika** [comment]
There are a number of factors at play here, although to get the desired style may actual be a simple change to the Mapfile. 

The "merging" of lines for roads (sometimes called road casing) is handled by using feature caching in the MapServer rendering code. When the cache is used all features are drawn first using a single style, and then subsequent styles are drawn in a next step using the feature from the cache. When the cache is not used then the "painter's model" is used of drawing each style on each feature in turn, producing the unwanted outlines in the images above. 

The first (undocumented - although I'm not sure where it could be documented) issue is that caching is only used when there are attribute bindings for the first STYLE in a CLASS. In this case there are two styles used, each with its own attribute bindings. Caching was disabled for this case in #3976. There was a follow-up issue #3995 to allow caching in scenarios such as this, but it was never implemented. 

In this use-case the features are all bound to the same values, so in initial tests turning on caching (using a new PROCESSING directive - `USE_DRAWING_CACHE`) produce the desired results. This could be added to MapServer if thought beneficial. 

An alternative workaround is to make use of the STYLE [OUTLINEWIDTH](https://www.mapserver.org/mapfile/style.html#mapfile-style-outlinewidth) introduced in [RFC 49](https://mapserver.org/development/rfc/ms-rfc-49.html#outlinewidth-on-line-layers) for this use case. 

The `OUTLINEWIDTH` is always in pixels, and is not dependent on scale. In this example the width is always 2 greater than the width of the line representing the road. The use of 2 styles could therefore be rewritten as follows:

```
CLASS
	EXPRESSION ('[type]' = 'motorway')
	STYLE
		WIDTH [width_motorway_in]
		MAXWIDTH 300
		COLOR "#ffbe5a"
                OUTLINECOLOR "#969696"
                OUTLINEWIDTH 2
	END
END
```

This avoids the issue of multiple bindings and multiple styles and uses the caching feature to produce road casing. 

![image](https://github.com/MapServer/MapServer/assets/490840/f4b47da1-03da-440c-952f-743f8f2ae172)

Unfortunately it does not fully resolve the issue. Whilst debugging there was a strange anomaly where attempting to use an attribute for `WIDTH` would produce different a different image than when the same `WIDTH` was hard-coded. On further investigations the `OUTLINEWIDTH` is calculated before binding the shape attribute to the style, and so the default width of 1 is always used. I'll create a separate issue for this. 

One possible option is to make use of [SYMBOLSCALEDENOM](https://www.mapserver.org/mapfile/layer.html#mapfile-layer-symbolscaledenom) on the layer to scale the width of the line automatically rather than using scaletokens.

**Comment by geographika** [comment]
Fixed with #6964
